crates/sui-core/src/authority/consensus_tx_status_cache.rs

```diff
@@ -0,0 +1,316 @@
+        self.status_notify_read
+            .notify(&transaction_position, &status);
+    }
+
+    pub async fn notify_read_transaction_status(
+        &self,
+        transaction_position: ConsensusTxPosition,
+        old_status: Option<ConsensusTxStatus>,
+    ) -> NotifyReadConsensusTxStatusResult {
+        let registration = self.status_notify_read.register_one(&transaction_position);
+        {
+            let inner = self.inner.read();
+            if let Some(status) = inner.transaction_status.get(&transaction_position) {
+                if Some(status) != old_status.as_ref() {
+                    if let Some(old_status) = old_status {
+                        // The only scenario where the status may change, is when the transaction
+                        // is initially fastpath certified, and then later finalized or rejected.
+                        assert_eq!(old_status, ConsensusTxStatus::FastpathCertified);
+                    }
+                    return NotifyReadConsensusTxStatusResult::Status(*status);
+                }
+            }
+            // Inner read lock dropped here.
+        }
+
+        let expiration_check = async {
+            loop {
+                {
+                    let inner = self.inner.read();
+                    if let Some(last_committed_round) = inner.last_committed_round {
+                        if transaction_position.block.round as u64 + self.gc_depth
+                            < last_committed_round
+                        {
+                            return last_committed_round;
+                        }
+                    }
+                }
+                sleep(Duration::from_millis(50)).await;
+            }
+        };
+        tokio::select! {
+            status = registration => NotifyReadConsensusTxStatusResult::Status(status),
+            last_committed_round = expiration_check => NotifyReadConsensusTxStatusResult::Expired(last_committed_round),
+        }
+    }
+
+    pub async fn update_last_committed_round(&self, round: u64) {
+        debug!("Updating last committed round: {}", round);
+        let mut inner = self.inner.write();
+        while let Some(&next_round) = inner.round_lookup_map.keys().next() {
+            if next_round + self.gc_depth < round {
+                let transactions = inner.round_lookup_map.remove(&next_round).unwrap();
+                for tx in transactions {
+                    inner.transaction_status.remove(&tx);
+                }
+            } else {
+                break;
+            }
+        }
+        inner.last_committed_round = Some(round);
+    }
+
+    /// Returns true if the position is too far ahead of the last committed round.
+    pub fn check_position_too_ahead(&self, position: &ConsensusTxPosition) -> SuiResult<()> {
+        let inner = self.inner.read();
+        if let Some(last_committed_round) = inner.last_committed_round {
+            // TODO(fastpath): Is using gc_depth appropriate here?
```

Comment:
Most probably `gc_depth` doesn't make too much sense here, but equally it simplifies a bit how we are thinking about thresholds. So it might be ok for now instead of introducing a separate parameter.